### PR TITLE
BUG: Fix unexpected sort in groupby

### LINKED ISSUE #17537: BUG: df.groupby(sort=False) sorts multi-index-frames
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame([
        [4, 2, 'x'],
        [3, 1, 'y'],
    ],
    columns=['A','B','C']).set_index(['A', 'B'])

print(df)
# Consider this DataFrame:
#
# >       C
# > A B
# > 4 2  x
# > 3 1  y

# Iterating over the group works if both levels
# of the multi-index are used for grouping.
for idx, group in df.groupby(level=[0, 1], sort=False):
    print(idx)
# > (4, 2)
# > (3, 1)

# However, grouping by only one level,
# suddenly sorts the index.
for idx, group in df.groupby(level=0, sort=False):
    print(idx)
# > 3
# > 4

# If the DataFrame has only one index,
# it works correctly
df2 = pd.DataFrame([
        [4, 2, 'x'],
        [3, 1, 'y'],
    ],
    columns=['A','B','C']).set_index(['A'])

print(df2)
# >     B  C
# > A
# > 4  2  x
# > 3  1  y

for idx, group in df2.groupby(level=0, sort=False):
    print(idx)
# > 4
# > 3
```
#### Problem description

DataFrame.groupby() has a parameter that selects whether the result should be sorted by groups.

However, if the DataFrame has a multi-index and the grouping is only done by one index, the result is sorted regardless of the value of `sort`.
Grouping by more than one index works.
Passing the single index as a list `[0]` does not fix the problem.

#### Expected Output
```python
for idx, group in df.groupby(level=0, sort=False):
    print(idx)
```
should yield
```
4
3
```

#### Output of ``pd.show_versions()``
<details>
INSTALLED VERSIONS

commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-33-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: 
LANG: en_GB.UTF-8
LOCALE: de_DE.UTF-8

pandas: 0.21.0.dev+450.g6eadb87fe
pytest: None
pip: 9.0.1
setuptools: 36.0.1
Cython: None
numpy: 1.13.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #17537
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -2613,6 +2613,13 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
 
             level = None
             key = group_axis
+        elif key is None:
```
Comment (jreback) [inline pandas/core/groupby.py]: hate adding logic here, this function is already impenetrable, can you incorporate this to existing?

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -1813,15 +1813,15 @@ def test_groupby_level(self):
...
-        result1 = frame.groupby(level='second').sum()
+        result0 = frame.groupby(level='first', sort=False).sum()
+        result1 = frame.groupby(level='second', sort=False).sum()
         assert_frame_equal(result0, expected0)
         assert_frame_equal(result1, expected1)
 
         # axis=1
 
-        result0 = frame.T.groupby(level=0, axis=1).sum()
-        result1 = frame.T.groupby(level=1, axis=1).sum()
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: for a couple of these that you changed can you also add the sort=True case (maybe parametrize on sort=)?

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note

**Comment by Licht-T** [comment]
@jreback Thanks for your review.
- Changed the location of the solution to improve readability
- Parameterized `sort` parameter of `groupby` in tests

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -2626,6 +2626,14 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     elif isinstance(key, BaseGrouper):
         return key, [], obj
 
+    if key is None and isinstance(group_axis, MultiIndex):
```
Comment (jreback) [inline pandas/core/groupby.py]: can you simplify this. e.g. maybe put this coercion higher (before the giant if/then)

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -1791,18 +1791,19 @@ def aggfun(ser):
         agged2 = df.groupby(keys).aggregate(aggfun)
         assert len(agged2.columns) + 1 == len(df.columns)
 
-    def test_groupby_level(self):
+    @pytest.mark.parametrize('sort', [True, False])
+    def test_groupby_level(self, sort):
         frame = self.mframe
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: add the issue number here as well as a comment

**Comment by Licht-T** [comment]
@jreback Thank you for comments. Fixed.

File: pandas/core/groupby.py [hunk 5]
```diff
@@ -2586,6 +2586,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     """
     group_axis = obj._get_axis(axis)
 
+    if key is None and level is not None and \
+       isinstance(group_axis, MultiIndex):
```
Comment (jreback) [inline pandas/core/groupby.py]: so maybe move this down a bit (under ``level is not None``). I don't want this to be a bespoke condition. I *think* you can remove the isinstance check of MultiIndex.

File: pandas/core/groupby.py [hunk 6]
```diff
@@ -2586,6 +2586,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
+        if is_list_like(level) and len(level) == 1:
+            level = level[0]
+
+        if is_scalar(level):
+            key = group_axis.get_level_values(level)
```
Comment (jreback) [inline pandas/core/groupby.py]: put a comment on what is going on here. *maybe* we can incorporate this below as well.

I am trying to remove as many special cases as possible.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.21.0.txt [hunk 7]
```diff
@@ -538,6 +538,7 @@ Groupby/Resample/Rolling
 - Bug in ``Series.resample(...).apply()`` where an empty ``Series`` modified the source index and did not return the name of a ``Series`` (:issue:`14313`)
 - Bug in ``.rolling(...).apply(...)`` with a ``DataFrame`` with a ``DatetimeIndex``, a ``window`` of a timedelta-convertible and ``min_periods >= 1` (:issue:`15305`)
 - Bug in ``DataFrame.groupby`` where index and column keys were not recognized correctly when the number of keys equaled the number of elements on the groupby axis (:issue:`16859`)
+- Bug in ``DataFrame.groupby`` where the single level selection from ``MultiIndex`` occurs unexpected index sorting (:issue:`17537`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: "occurs" -> "incurs"? Or maybe "causes"?

**Review by TomAugspurger — COMMENTED** [review]
@Licht-T this looks pretty close. Could you fixup the merge conflict and take a look at @jreback's comments in the next couple days?

File: doc/source/whatsnew/v0.21.0.txt (hunk 7 above)
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: where a single level selection from a ``MultiIndex`` unexpectedly sorts.

File: pandas/core/groupby.py [hunk 8]
```diff
@@ -2586,10 +2586,21 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     """
     group_axis = obj._get_axis(axis)
 
-    # validate that the passed level is compatible with the passed
+    # validate that the passed single level is compatible with the passed
     # axis of the object
     if level is not None:
-        if not isinstance(group_axis, MultiIndex):
+        # TODO: These two conditions are almost same.
```
Comment (jreback) [inline pandas/core/groupby.py]: ok for now. can you come back in a future PR and see what we can do with all the conditions in this section. getting pretty unweildy (and document as much as possible).

File: pandas/core/groupby.py [hunk 9]
```diff
@@ -2586,10 +2586,21 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
+        # We should combine two.
+        if isinstance(group_axis, MultiIndex):
+            if is_list_like(level) and len(level) == 1:
```
Comment (jreback) [inline pandas/core/groupby.py]: actually this condition I think you can pull out of the MultiIndex check here (as the else is the same condition)

File: pandas/core/groupby.py (hunk 8 above)
Comment (Licht-T) [reply pandas/core/groupby.py]: @jreback Okay. I'll do that. These are too complicated to do refactoring in this PR, I think.

File: pandas/core/groupby.py (hunk 9 above)
Comment (Licht-T) [reply pandas/core/groupby.py]: @jreback I am aware of this, but it seems that there are some processes only for non-`MultiIndex` condition in `else`. We have to consider carefully whether these are applicable for `MultiIndex`.
https://github.com/pandas-dev/pandas/pull/17621/files/e4cdd0726e685b0216056ba224ed363bf1e836f9#diff-720d374f1a709d0075a1f0a02445cd65R2618

Comment (Licht-T) [reply pandas/core/groupby.py]: When these are applicable, we also have to check if there is no side effect to subsequent processes.

**Comment by jreback** [comment]
can you rebase

**Comment by Licht-T** [comment]
@jreback Rebased.

**Comment by jreback** [comment]
@Licht-T can you rebase and push once again, want to get all green here.

**Comment by Licht-T** [comment]
@jreback Now all green!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Licht-T

**Pull request merged**